internal/util/participant_utils.go

```diff
@@ -40,38 +40,38 @@ func ContainsAll(haystack, needles []*drand.Participant) bool {
 	return true
 }
 
+// Without removes needle from the haystack. Careful: it modifies the input slice but also returns the resulting slice.
+// It removes all instances of needle, and zeros the removed items to allow garbage collection.
 func Without(haystack []*drand.Participant, needle *drand.Participant) []*drand.Participant {
 	if haystack == nil {
 		return nil
 	}
 
-	indexToRemove := -1
-	for i, v := range haystack {
+	// ret will reuse the underlying array of haystack
+	ret := haystack[:0]
```

Comment:
fiddling with this in the go playground, I'm not entirely surely it does what is expected.

https://go.dev/play/p/wBc1tJYq4Ra

the `ret` value here gets a nil, but misses the last appended value because they share an underlying array

Reply:
the tests may work because the needle is always at the end - can you add a test with the needle in the middle to see if it still works as expected?

Reply:
The ret doesn't misses the last appended value, it re-uses the storage of the underlying array, so that last appended value gets overwritten when you nilify it.
https://go.dev/play/p/VkqjnCl-7nb

But sure, I can add more tests.

Reply:
ahh I understand now - it trims _from_ len(ret) onwards